src/stream/src/cache/range_lru.rs

```diff
@@ -0,0 +1,542 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::BTreeMap;
+use std::ops::Bound::{self, *};
+
+use risingwave_common::types::ScalarImpl;
+
+type ScalarRange = (Bound<ScalarImpl>, Bound<ScalarImpl>);
+
+#[derive(Default)]
+struct RangeLru {
```

Comment:
Not necessary, but I guess S: Scalar may be better for performance.

Reply:
> Not necessary, but I guess S: Scalar may be better for performance.

Do we have a similar trait for `Datum` or `Row`? 😢